### ISSUE TITLE
Query: parameter nullability sniffing broken for global query filters

### ISSUE THREAD

**Issue by bachratyg** [dotnet/efcore#24645]
I was under the impression that as of #15892 global query filters are subject to the same parameter nullability optimizations as the rest of the query however this does not seem to be the case for some expressions.

### repro code
This code is a simplification of a multitenant system where most actors are allowed to access their own tenant only, while a very few privileged actors are allowed to view all tenants.

``` xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.5" />
  </ItemGroup>
</Project>
```
```C#
class BlogContext : DbContext
{
	public DbSet<Blog> Blogs { get; set; }
	// For sake of simplicity this is a basic property
	// In real project this comes from a constructor-injected service
	public string TenantId { get; set; }
	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
	{
		optionsBuilder
			.EnableDetailedErrors()
			.EnableSensitiveDataLogging()
			// Not particularly important, repro code won't hit the db, same bug with Oracle
			.UseSqlServer("Data Source=localhost;Integrated Security=true", sql => sql.UseRelationalNulls());
	}
	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		modelBuilder
			.Entity<Blog>()
			// For normal users the tenant id is not null
			// For privileged users the tenant id is null
			.HasQueryFilter(b => b.TenantId == this.TenantId || this.TenantId == null);
	}
}
class Blog
{
	public int Id { get; set; }
	public int SomeValue { get; set; }
	[Required] public string TenantId { get; set; }
}
class Program
{
	static void Main()
	{
		using var db = new BlogContext();
		db.TenantId = "asd";
		var q = db.Blogs.Where(b => b.SomeValue == 1);
		Console.WriteLine(q.ToQueryString());
	}
}
```

Expected result (when tenant id is not null)
``` SQL
SELECT [b].[Id], [b].[SomeValue], [b].[TenantId]
FROM [Blogs] AS [b]
WHERE ([b].[TenantId] = @__ef_filter__TenantId_0) AND ([b].[SomeValue] = 1)
```

Actual result
``` SQL
SELECT [b].[Id], [b].[SomeValue], [b].[TenantId]
FROM [Blogs] AS [b]
WHERE (([b].[TenantId] = @__ef_filter__TenantId_0) OR (@__ef_filter__p_1 = CAST(1 AS bit))) AND ([b].[SomeValue] = 1)
```
The generated OR expression prevents the database from using the index on TenantId. This practically kills perf for most queries since the tenant is the first column in most indexes.

Some nullability sniffing is still going on: when the tenant is null the first part of the OR expression is optimized away:
``` SQL
SELECT [b].[Id], [b].[SomeValue], [b].[TenantId]
FROM [Blogs] AS [b]
WHERE (@__ef_filter__p_1 = CAST(1 AS bit)) AND ([b].[SomeValue] = 1)
```
It seems like `@tenantId IS NULL` gets prematurely optimized to `@tenantIdIsNull = 1` which then could not be optimized further since the new parameter is always non-null.

#### Use direct filters
This gets optimized properly:
``` C#
var q = db.Blogs.Where(b => b.TenantId == db.TenantId || db.TenantId == null).Where(b => b.Id == 1);
```
Pro: no hidden magic
Pro: query caching and db indexing works as expected
Con: litters the code with filters, easy to forget that filtering needs to be applied
Con: difficult to filter navigations, easy to make a mistake
Con: necessary tenant information needs to be injected to all query callsites
Con: why do we even have global query filters?

#### Always assume tenant and use IgnoreQueryFilters where necessary

``` C#
var q = (tenantId != null ? db.Blogs : db.Blogs.IgnoreQueryFilters()).Where(b => b.Id == 1);
```
Pro: the filter logic is defined once at the model level
Pro: query caching and db indexing works as expected
Con: bypasses all query filters not just the tenant filter.
Con: litters the code with conditional constructs 
Con: necessary tenant information needs to be injected to all query callsites

#### Use different contexts

derived from the same base context but with different global filters

Pro: query caching and db indexing works as expected
Con: needs some plumbing i.e. a context factory that returns this or that context based on tenant information
Con: does not scale, needs a separate context for all combinations when multiple filters are affected
Con: separate model, query cache, etc creates perf issues on the client

#### RelationalParameterBasedSqlProcessor

Based on https://github.com/dotnet/efcore/issues/17598#issuecomment-691354789
Basically a custom RelationalParameterBasedSqlProcessor is injected that evaluates expressions like `@p = constant` and `term OR constant` based on parameter values.

Pro: business code is clean
Pro: db indexing works as expected
Con: bypasses the command cache leading to some perf issues on the client, 

### environment
EF Core version: 5.0.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0
Operating system: Windows 10 20H2 (19042.906)
IDE: Visual Studio 2019 16.9.3

**Comment by AndriySvyryd** [comment]
**EF Triage:** Currently this is by-design and the `RelationalParameterBasedSqlProcessor` workaround is the best option. After https://github.com/dotnet/efcore/issues/17598 is implemented there might be an easier/better workaround.
